lua/frecency/picker.lua

```diff
@@ -0,0 +1,264 @@
+        o.updated_finder = finders.new_table { results = M.results, entry_maker = M.maker }
+      end
+      o.prompt = matched and query_text:sub(matched:len() + 1) or query_text
+      return o
+    end,
+    attach_mappings = function(prompt_bufnr)
+      actions.select_default:replace_if(function()
+        return vim.fn.complete_info().pum_visible == 1
+      end, function()
+        local keys = vim.fn.complete_info().selected == -1 and "<C-e><Bs><Right>" or "<C-y><Right>:"
+        local accept_completion = vim.api.nvim_replace_termcodes(keys, true, false, true)
+        vim.api.nvim_feedkeys(accept_completion, "n", true)
+      end)
+      return true
+    end,
+  })
+  M.picker:find()
+  util.buf_set {
+    M.picker.prompt_bufnr,
+    options = { filetype = "frecency", completefunc = "frecency#FrecencyComplete" },
+    mappings = {
+      expr = true,
+      ["i|<Tab>"] = "pumvisible() ? '<C-n>'  : '<C-x><C-u>'",
+      ["i|<S-Tab>"] = "pumvisible() ? '<C-p>'  : ''",
+    },
+  }
+end
+
+---Update Frecency Picker result
+---@param filter string
+---@return boolean
+---@TODO: make it more readable please :D
+M.update_results = function(filter)
+  local filter_updated = false
+  local ws_dir = filter and M.config.user_workspaces[filter]
+
+  if filter == "LSP" and not vim.tbl_isempty(M.lsp_workspaces) then
+    ws_dir = M.lsp_workspaces[1]
+  elseif filter == "CWD" then
+    ws_dir = M.cwd
+  end
+
+  if ws_dir ~= M.active_filter then
+    filter_updated = true
+    M.active_filter, M.active_filter_tag = ws_dir, filter
+  end
+
+  if vim.tbl_isempty(M.results) or filter_updated then
+    M.results = M.db.get_files { ws_dir = ws_dir, show_unindexed = M.config.show_unindexed, with_score = true }
+  end
+  return filter_updated
+end
+
+---Create entry maker function.
+---@param entry table
+---@return function
+---FIXME: path transacted with icons being pre-appended instead of appended
+M.maker = function(entry)
+  local items = {}
+  if M.config.show_scores then
+    table.insert(items, { width = 8 })
+  end
+
+  if M.config.show_filter_column then
+    local width = 0
+    -- TODO: Only add +1 if M.show_filter_thing is true, +1 is for the trailing slash
+    if M.active_filter and M.active_filter_tag == "LSP" then
+      width = #(utils.path_tail(M.active_filter)) + 1
+    elseif M.active_filter then
+      width = #(Path:new(M.active_filter):make_relative(os_home)) + 1
+    end
+    table.insert(items, { width = width })
+  end
+
+  if not M.config.disable_devicons and has_devicons then
+    table.insert(items, { width = 2 }) -- icon column
+  end
+
+  table.insert(items, { remaining = true })
+
+  return {
+    filename = entry.path,
+    ordinal = entry.path,
+    name = entry.path,
+    score = entry.score,
+    display = function(e)
+      return entry_display.create {
+        separator = "",
+        items = items,
+        hl_chars = { [os_path_sep] = "TelescopePathSeparator" },
+      }(M.display_items(e))
```

Comment:
This is not good. This would create a new entry displayer for each entry which would just add a lot of complexity.
The idea is to create a entry_displayer onces and then run this displayer for n entries

Reply:
😣 this the part I took the longest to refactor.

I'll tag you when a refactor it further but basically you are saying that I should have two separate functions?